File: paper/src/main/java/net/skullian/skyfactions/paper/SkyLoader.java
package net.skullian.skyfactions.paper;

import io.papermc.paper.plugin.loader.PluginClasspathBuilder;
import io.papermc.paper.plugin.loader.PluginLoader;
import io.papermc.paper.plugin.loader.library.impl.MavenLibraryResolver;
import net.skullian.skyfactions.common.util.LibraryParser;
import org.eclipse.aether.artifact.DefaultArtifact;
import org.eclipse.aether.graph.Dependency;
import org.eclipse.aether.repository.RemoteRepository;

import java.io.IOException;
import java.net.URISyntaxException;

/**
 * The dependency loader for the SkyFactions plugin.
 * To avoid large plugin sizes, we use paper's (experimental) dependency loader to load
 * dependencies during runtime.
 */
@SuppressWarnings("UnstableApiUsage")
public class SkyLoader implements PluginLoader {

    private final LibraryParser parser = new LibraryParser();

    @Override
    public void classloader(PluginClasspathBuilder classpathBuilder) {

        try {
            LibraryParser.LibraryData data = parser.parseLibraries();
            MavenLibraryResolver resolver = new MavenLibraryResolver();

            data.repositories().forEach(repository -> {
                resolver.addRepository(
                        new RemoteRepository.Builder(repository.id(), repository.type(), repository.url()).build()
                );
            });

            data.artifacts().forEach(library -> {
                    resolver.addDependency(
                            new Dependency(new DefaultArtifact(String.format("%s:%s:%s", library.groupId(), library.artifactId(), library.version())), null)
                    );
            });

            classpathBuilder.addLibrary(resolver);
        } catch (IOException | URISyntaxException e) {
            throw new RuntimeException(e);
        }

    }
}

File: paper/src/main/kotlin/net/skullian/skyfactions/paper/SkyLoader.kt
package net.skullian.skyfactions.paper

import io.papermc.paper.plugin.loader.PluginClasspathBuilder
import io.papermc.paper.plugin.loader.PluginLoader
import io.papermc.paper.plugin.loader.library.impl.MavenLibraryResolver
import net.skullian.skyfactions.common.util.LibraryParser
import org.eclipse.aether.artifact.DefaultArtifact
import org.eclipse.aether.graph.Dependency
import org.eclipse.aether.repository.RemoteRepository

/**
 * The dependency loader for the SkyFactions plugin.
 * To avoid large plugin sizes, we use paper's (experimental) dependency loader to load
 * dependencies during runtime.
 */
@Suppress("UnstableApiUsage")
class SkyLoader : PluginLoader {

    private val parser = LibraryParser()

    override fun classloader(classpathBuilder: PluginClasspathBuilder) {

        val libraryData = parser.parseLibraries()
        val resolver = MavenLibraryResolver()

        libraryData.repositories.forEach { repository ->
            resolver.addRepository(
                RemoteRepository.Builder(repository.name, repository.type, repository.url).build()
            )
        }

        libraryData.artifacts.forEach { library ->
            resolver.addDependency(
                Dependency(DefaultArtifact("${library.groupId}:${library.artifactId}:${library.version}"), null)
            )
        }

        classpathBuilder.addLibrary(resolver)
    }
}